### PR TITLE
countperl: Add JSON format output capability

### PR DESCRIPTION
**Pull request by polamjag**
Hello! 

I implemented option to output the analysis results in JSON.

I think it's important to output analysis results in a machine-readable format because we become able to easily process and analyze results in continuous integrations and so on.

Example result:

```
% bin/countperl lib --json
{"statistics":{"mccabe_complexity":{"code_not_in_any_subroutine":{"standard_deviation":"0.00","median":"1.00","mean":"1.00","min":1,"max":1},"sub_complexity":{"median":"2.00","standard_deviation":"3.73","mean":"3.18","min":1,"max":20}},"counts":{"lines_of_non_sub_code":114,"packages_found":3,"total_code_lines":614,"subs_and_methods_count":50},"subroutine_sizes":{"min":3,"max":42,"mean":"10.00","median":"8.00","standard_deviation":"8.40"},"file_count":3},"subs":[{"lines":38,"path":"lib/Perl/Metrics/Simple/Analysis/File.pm","mccabe_complexity":20,"name":"_rewrite_moose_method_modifiers"},{"path":"lib/Perl/Metrics/Simple/Analysis/File.pm","lines":20,"mccabe_complexity":14,"name":"get_node_length"},{"mccabe_complexity":13,"name":"_init","lines":42,"path":"lib/Perl/Metrics/Simple/Analysis/File.pm"},{"lines":20,"path":"lib/Perl/Metrics/Simple/Analysis/File.pm","mccabe_complexity":10,"name":"measure_complexity"},{"lines":13,"path":"lib/Perl/Metrics/Simple.pm","name":"analyze_files","mccabe_complexity":6},{"path":"lib/Perl/Metrics/Simple.pm","lines":11,"name":"is_perl_file","mccabe_complexity":6},{"lines":12,"path":"lib/Perl/Metrics/Simple.pm","name":"_has_perl_shebang","mccabe_complexity":6},{"path":"lib/Perl/Metrics/Simple/Analysis.pm","lines":8,"mccabe_complexity":5,"name":"is_ref"},{"path":"lib/Perl/Metrics/Simple/Analysis/File.pm","lines":16,"mccabe_complexity":5,"name":"is_hash_key"},{"name":"find_files","mccabe_complexity":4,"lines":10,"path":"lib/Perl/Metrics/Simple.pm"},{"name":"_get_min_max_values","mccabe_complexity":4,"path":"lib/Perl/Metrics/Simple/Analysis.pm","lines":11},{"path":"lib/Perl/Metrics/Simple/Analysis/File.pm","lines":23,"mccabe_complexity":4,"name":"analyze_main"},{"name":"_get_packages","mccabe_complexity":4,"lines":14,"path":"lib/Perl/Metrics/Simple/Analysis/File.pm"},{"path":"lib/Perl/Metrics/Simple/Analysis/File.pm","lines":18,"name":"_iterate_over_subs","mccabe_complexity":4},{"lines":14,"path":"lib/Perl/Metrics/Simple.pm","mccabe_complexity":3,"name":"list_perl_files"},{"lines":10,"path":"lib/Perl/Metrics/Simple/Analysis.pm","name":"new","mccabe_complexity":3},{"mccabe_complexity":3,"name":"_get_mean_median_std_dev","lines":12,"path":"lib/Perl/Metrics/Simple/Analysis.pm"},{"path":"lib/Perl/Metrics/Simple/Analysis/File.pm","lines":12,"mccabe_complexity":3,"name":"_create_ppi_document"},{"path":"lib/Perl/Metrics/Simple/Analysis/File.pm","lines":10,"mccabe_complexity":3,"name":"_prune_non_code_lines"},{"path":"lib/Perl/Metrics/Simple.pm","lines":5,"mccabe_complexity":2,"name":"should_be_skipped"},{"lines":3,"path":"lib/Perl/Metrics/Simple/Analysis.pm","name":"_numerically","mccabe_complexity":2},{"path":"lib/Perl/Metrics/Simple/Analysis.pm","lines":29,"mccabe_complexity":2,"name":"_init"},{"mccabe_complexity":2,"name":"_summary_stats_main_complexity","path":"lib/Perl/Metrics/Simple/Analysis.pm","lines":10},{"mccabe_complexity":2,"name":"logic_keywords","lines":4,"path":"lib/Perl/Metrics/Simple/Analysis/File.pm"},{"name":"logic_operators","mccabe_complexity":2,"path":"lib/Perl/Metrics/Simple/Analysis/File.pm","lines":4},{"mccabe_complexity":2,"name":"method_modifiers","path":"lib/Perl/Metrics/Simple/Analysis/File.pm","lines":4},{"path":"lib/Perl/Metrics/Simple/Analysis/File.pm","lines":4,"name":"hashify","mccabe_complexity":2},{"mccabe_complexity":1,"name":"new","path":"lib/Perl/Metrics/Simple.pm","lines":6},{"path":"lib/Perl/Metrics/Simple/Analysis.pm","lines":4,"mccabe_complexity":1,"name":"files"},{"mccabe_complexity":1,"name":"data","path":"lib/Perl/Metrics/Simple/Analysis.pm","lines":4},{"mccabe_complexity":1,"name":"file_count","path":"lib/Perl/Metrics/Simple/Analysis.pm","lines":4},{"name":"lines","mccabe_complexity":1,"lines":4,"path":"lib/Perl/Metrics/Simple/Analysis.pm"},{"name":"packages","mccabe_complexity":1,"path":"lib/Perl/Metrics/Simple/Analysis.pm","lines":4},{"lines":4,"path":"lib/Perl/Metrics/Simple/Analysis.pm","name":"package_count","mccabe_complexity":1},{"lines":4,"path":"lib/Perl/Metrics/Simple/Analysis.pm","mccabe_complexity":1,"name":"file_stats"},{"lines":4,"path":"lib/Perl/Metrics/Simple/Analysis.pm","mccabe_complexity":1,"name":"main_stats"},{"lines":4,"path":"lib/Perl/Metrics/Simple/Analysis.pm","name":"summary_stats","mccabe_complexity":1},{"path":"lib/Perl/Metrics/Simple/Analysis.pm","lines":4,"name":"subs","mccabe_complexity":1},{"mccabe_complexity":1,"name":"sub_count","lines":4,"path":"lib/Perl/Metrics/Simple/Analysis.pm"},{"path":"lib/Perl/Metrics/Simple/Analysis.pm","lines":9,"name":"_make_summary_stats","mccabe_complexity":1},{"path":"lib/Perl/Metrics/Simple/Analysis.pm","lines":9,"name":"_summary_stats_sub_length","mccabe_complexity":1},{"name":"_summary_stats_sub_complexity","mccabe_complexity":1,"path":"lib/Perl/Metrics/Simple/Analysis.pm","lines":9},{"path":"lib/Perl/Metrics/Simple/Analysis/File.pm","lines":7,"mccabe_complexity":1,"name":"new"},{"name":"_make_pruned_document","mccabe_complexity":1,"lines":8,"path":"lib/Perl/Metrics/Simple/Analysis/File.pm"},{"name":"all_counts","mccabe_complexity":1,"path":"lib/Perl/Metrics/Simple/Analysis/File.pm","lines":11},{"mccabe_complexity":1,"name":"path","path":"lib/Perl/Metrics/Simple/Analysis/File.pm","lines":4},{"lines":4,"path":"lib/Perl/Metrics/Simple/Analysis/File.pm","mccabe_complexity":1,"name":"main_stats"},{"name":"subs","mccabe_complexity":1,"path":"lib/Perl/Metrics/Simple/Analysis/File.pm","lines":4},{"path":"lib/Perl/Metrics/Simple/Analysis/File.pm","lines":4,"name":"packages","mccabe_complexity":1},{"mccabe_complexity":1,"name":"lines","lines":4,"path":"lib/Perl/Metrics/Simple/Analysis/File.pm"},{"path":"lib/Perl/Metrics/Simple.pm","lines":21,"name":"{code not in named subroutines}","mccabe_complexity":1},{"mccabe_complexity":1,"name":"{code not in named subroutines}","path":"lib/Perl/Metrics/Simple/Analysis.pm","lines":20},{"path":"lib/Perl/Metrics/Simple/Analysis/File.pm","lines":73,"mccabe_complexity":1,"name":"{code not in named subroutines}"}]}
```

### CODE REVIEW

**Comment by matisse** [comment]
Thank you very much!
I will have a closer look in the next couple of days.

File: bin/countperl [hunk 1]
```diff
@@ -7,6 +7,7 @@ use Carp qw();
 use Readonly;
 use Pod::Usage qw(pod2usage);
 use Perl::Metrics::Simple qw(0.13);
+use JSON::PP qw(encode_json);
```
Comment (matisse) [inline bin/countperl:10]: I'll need to add this to Build.PL and Makefile.PL
(I also need to update the `Changes` file)

File: bin/countperl [hunk 2]
```diff
@@ -482,6 +490,72 @@ sub get_class_by_count {
 
 ### /HTML
 
+### JSON
```
Comment (matisse) [inline bin/countperl:493]: I should consider moving the JSON (and HTML) stuff into modules in `lib` so the functions are available outside of `countperl`, or maybe make `countperl` testable, in either case I could add some test coverage.

**Review by matisse — APPROVED** [review]
Thanks for doing this.

I'm going to merge this into a branch in my own repository where I'll do some additional work on it before eventually releasing it.

**Comment by polamjag** [comment]
Thanks! Let me know if there's anything I can do to help.

**Comment by matisse** [comment]
Sorry I have had some extra work come up and am delayed - have not forgotten.

- Matisse

**Comment by matisse** [comment]
FYI:

I uploaded Perl-Metrics-Simple-0.19 today.

-Matisse

**Pull request merged**